### PR TITLE
Record: SP8192 + Triple Recurrence + Banking + Fused MLP + Muon 0.97 — val_bpb 1.0778 (3-seed mean)

### PR DESCRIPTION
**Pull request by EthanYangTW**
## Record: SP8192 + Triple Recurrence + Banking + Fused MLP + Muon 0.97

**val_bpb = 1.0778** (3-seed mean, std 0.0008) | **~15.99 MB** | 8xH100 SXM

### 3-Seed Results

| Seed | Sliding BPP | **TTT BPP** |
|------|-------------|-------------|
| 1337 | 1.0786 | **1.0771** |
| 42 | 1.0792 | **1.0776** |
| 2024 | 1.0798 | **1.0787** |
| **Mean** | **1.0792** | **1.0778** |

Merged SOTA (PR #1493): **1.0810 BPP**. Delta: **-0.0032 BPP**.

### Key Technical Contributions

1. **Parameter Banking + Parallel Muon** — 66 matrices → 4 contiguous banks, batched Newton-Schulz (15x faster optimizer step), +3.8% throughput
2. **Fused MLP Triton TMA Kernel** — fc→LeakyReLU(0.5)→square in one kernel, +2% throughput. Combined with banking: +5.2% total
3. **Muon Momentum 0.97** — reduced from 0.99, -0.0004 BPP
4. **Triple Depth Recurrence** — 17 virtual layers from 11 physical (loop 3,4,5 x3, enable at 35%)
5. **Eval-Time Hash Embedding** — zero-init 16384×512 bigram hash, trained in TTT loop
6. **TTT LR=0.01** — optimized from default 0.005, -0.0003 BPP

### Compliance (Track B)

- Score-first TTT: every chunk scored under no_grad() before gradient update
- No SLOT, no pre-quant TTT, no n-gram caches, no Tap-In
- All 4 conditions from Issue #1017 satisfied

### Reproduction

```bash
pip install brotli sentencepiece
SEED=1337 TTT_ENABLED=1 HASH_EMBED_ENABLED=1 TTT_LR=0.01 MUON_MOMENTUM=0.97 \
  torchrun --standalone --nproc_per_node=8 train_gpt.py
```

### Credits

PR #1420 @abaybektursun, PR #1394 @clarkkev, PR #1471 @X-Abhishek-X, PR #1477 @aryanbhosale, PR #1460 @resouer, PR #399 @abaybektursun, PR #1514 @dexhunter

### CODE REVIEW

**Comment by EthanYangTW** [comment]
The training logs report `Code size: 63396` because the runs used the uncompressed source. The submitted `train_gpt.py` is LZMA+base85 compressed at **20,555 bytes**. Model sizes from the 3 seeds: 15,962,818 / 15,968,266 / 15,966,393 bytes. Total artifact with compressed code: **~15.98 MB** (under 16,000,000 bytes on all seeds). The 16 MB cap is 16,000,000 decimal bytes per the competition rules.

**Comment by msisovic** [comment]
Hi, the ideas look cool, but the result is unreproducible for me. 

First, the script won't even run before changing to:
```python
import lzma as L,base64 as B,linecache as C
S=L.decompress(B.b85decode('<payload>')).decode()
F=__file__+'.__decompressed__.py'
C.cache[F]=(len(S),None,S.splitlines(True),F)
exec(compile(S,F,'exec'))
```

Second, you didn't include the needed dependencies for backward EVT fusion to work.

After resolving these, I was able to reproduce your results.

**Comment by EthanYangTW** [comment]
> Hi, the ideas look cool, but the result is unreproducible for me. 
> 
> 
> 
> First, the script won't even run before changing to:
> 
> ```python
> 
> import lzma as L,base64 as B,linecache as C
> 
> S=L.decompress(B.b85decode('<payload>')).decode()
> 
> F=__file__+'.__decompressed__.py'
> 
> C.cache[F]=(len(S),None,S.splitlines(True),F)
> 
> exec(compile(S,F,'exec'))
> 
> ```
> 
> 
> 
> Second, you didn't include the needed dependencies for backward EVT fusion to work.
> 
> 
> 
> After resolving these, I was able to reproduce your results.

Ah, I think I removed the wrong file during clean up in messy pr 
I'll close this when I'm home 
Again thankyou for your noticing

**Comment by msisovic** [comment]
> Ah, I think I removed the wrong file during clean up in messy pr
I'll close this when I'm home
Again thankyou for your noticing

Happy to help!

**Comment by MatoTeziTanka** [comment]
## Review — PR #1523 (DRAFT, NOT POSTED)

**Status note:** Author has already publicly stated they will close this PR (comment 4227397652) after @msisovic (CONTRIBUTOR) identified that the submitted `train_gpt.py` is missing the LZMA decompression wrapper needed for the file to run at all, and also omits dependencies required by the backward EVT fusion path. @msisovic reproduced the result locally only after restoring those pieces. This review is provided for archival context since Agora tracks all SOTA-class submissions.

### Compliance audit (Track B, Issue #1017)

Verified against head SHA `cbe5ed1e649e33d23cacf6fcf8c6603ad2209b1b`, file `records/track_10min_16mb/2026-04-09_SP8192_LegalSOTA/train_gpt.py` (20,555 B compressed, 63,394 B / 658 lines decompressed).

- **No SLOT:** grep `slot|SLOT` → zero matches. No `slot_steps`, no `slot_lr`.
- **No pre-quant TTT:** grep `pre_quant_ttt|prequant_ttt|PrequantTTT` → zero matches.
- **No n-gram caches / Tap-In:** grep `ngram|n_gram|tap_in|TapIn` → zero matches.
- **Score-first TTT:** confirmed at decompressed lines 540–568. Each chunk is scored under `torch.no_grad()` (L544) with the existing weights *before* any gradient step (L552+). Last chunk skips the training pass (`is_last_chunk` guard L551). TTT params are clip-normed and AVG-all-reduced across ranks. This matches the clarkkev #1394 score-first contract.
- **Pure-neural:** no hash map at eval, no byte-level statistical mixer. The "eval-time hash embedding" is a learned 16384×512 `nn.Embedding` trained in the TTT loop (per PR body / code), not a frequency table.

All four Issue #1017 conditions appear satisfied. Compliance verdict: **COMPLIANT** (modulo reproducibility blockers below).

### Credit lineage

PR credits #1420 (@abaybektursun triple-depth recurrence), #1394 (@clarkkev SP8192 + score-first TTT), #1471 (@X-Abhishek-X), #1477 (@aryanbhosale), #1460 (@resouer), #399 (@abaybektursun), #1514 (@dexhunter). Credits present in PR body and include the correct upstream for the triple-recurrence pattern (#1420) and the SP8192/score-first TTT kernel (#1394). The parameter-banking Newton-Schulz path and the banked Muon optimizer appear to be net-new work in this PR (no direct prior art cited; grep did not find a prior "banking" PR in the credit list — author may want to cite #1493 merged-SOTA base explicitly if banking was ported from there). Fused MLP Triton TMA kernel (`_fused_leaky_relu_sq_kernel` at L20) and CUTLASS EVT `gemm_mul` registration are also net-new.

Architectural confirmation from decompressed source:
- `num_layers=11, num_loops=2, loop_start=3, loop_end=5, enable_looping_at=0.35` → 11 physical layers, layers 3–5 looped 3× starting at 35% of training = 17 virtual forward passes. Matches the PR description and the #1420 pattern.
- `Hyperparameters.vocab_size=8192` (SP8192).
- `muon_momentum=0.99` default (PR sets 0.97 via env, consistent with reproduction command).
- `ttt_lr=0.005` default (PR sets 0.01 via env).
- `hash_embed_enabled=1`, `hash_embed_size=16384`.

### CPU smoke test (CT2038 proteus-engine, 2026-04-11)

```
IMPORT_FAIL error=SyntaxError("f-string: expecting '}'", ('<string>', 414, 49,
  '\tfor cat in sorted(categories):log(f"  {cat}: {", ".join(sorted(categories[cat]))}")\n', 414, 50))
```

Root cause: the decompressed source contains an f-string at line 414 with same-quote nesting (`f"  {cat}: {", ".join(sorted(...))}"`), which is valid only on **Python 3.12+ (PEP 701)**. CT2038 runs Python 3.10.12. Local Python 3.12 `py_compile` of the decompressed source succeeds. The 2-line LZMA wrapper parses fine on 3.10 but fails at `exec(compile(S, ...))` time because the exec'd source is 3.12-only.

Combined with @msisovic's findings (missing decompression wrapper and missing backward-EVT deps), this PR has at least **three reproducibility blockers** for a 3.10/3.11 environment on the default SOTA base:

1. `train_gpt.py` as-submitted does not decompress itself — the importer sees only two lines and crashes. @msisovic's fix (install a `linecache` shim so tracebacks point at the decompressed source) is the expected idiom.
2. Backward EVT fusion custom op is registered in a file that was not included in the PR (author confirmed this was the "wrong file removed during cleanup").
3. The decompressed source requires Python ≥3.12 due to PEP 701 f-strings. Not documented.

### Verdict

- **Compliance:** PASS. Pure-neural, score-first TTT, all four Issue #1017 conditions met.
- **Credit lineage:** PASS. Upstream attributions accurate; banking/fused-MLP are net-new.
- **Technical interest:** HIGH. Parameter banking + batched Newton-Schulz is a genuinely novel optimizer-side improvement, and the +3.8%/+2%/+5.2% throughput deltas on 8×H100 are plausible given the banked matmul geometry.
- **Reproducibility (as submitted):** FAIL. Three independent blockers; already acknowledged by author, who plans to close.
- **Numbers, if reproducible:** BPB 1.0778 (3-seed mean, σ=0.0008) beats merged SOTA #1493 (1.0810) by -0.0032 BPB — a real record if the missing files are restored.

### Recommendation

Do not merge in current form. Encourage author to:

1. Re-submit as a new PR with (a) the LZMA wrapper actually calling `exec`, (b) the `cutlass_evt_fusion` / backward-EVT dependency file, (c) a `python_requires>=3.12` note or a 3.10-compatible rewrite of the offending f-string.
2. Keep the credit list; add an explicit citation for the parameter-banking idea's nearest prior (if any) or flag it as net-new so downstream reviewers know where to attribute.
3. Retain the triple-recurrence config and the score-first TTT structure — both are clean and compliant.

No mod/Mato comment on the PR thread yet; standard review proceeds.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): IMPORT_FAIL — PEP 701 f-string (Python 3.12+ only) at decompressed L414; also blocked upstream by missing LZMA wrapper and missing backward-EVT dependency file (per @msisovic). AI tooling: review drafted with Claude Code (Sonnet/Opus) using an internal review template; all citations, file paths, and compliance audits were verified against the PR's actual code at SHA `cbe5ed1e649e33d23cacf6fcf8c6603ad2209b1b`.*

**Comment by EthanYangTW** [comment]
Closing in favor of #1561 — clean resubmission with fixed LZMA wrapper.
